man/recset.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.10.
.TH RECSET "1" "April 2014" "recset 1.7" "User Commands"
.SH NAME
recset \- manipulate fields in a recfile
.SH SYNOPSIS
.B recset
[\fIOPTION\fR]... [\fIFILE\fR]...
.SH DESCRIPTION
Alter or delete fields in records.
.TP
\fB\-\-no\-external\fR
don't use external descriptors.
.TP
\fB\-\-force\fR
alter the records even if violating record
restrictions.
.TP
\fB\-\-help\fR
print a help message and exit.
.TP
\fB\-\-version\fR
show version and exit.
.SS "Record selection options:"
.TP
\fB\-i\fR, \fB\-\-case\-insensitive\fR
make strings case\-insensitive in selection
expressions.
.TP
\fB\-t\fR, \fB\-\-type\fR=\fITYPE\fR
operate on records of the specified type only.
.TP
\fB\-e\fR, \fB\-\-expression\fR=\fIRECORD_EXPR\fR
selection expression.
.TP
\fB\-q\fR, \fB\-\-quick\fR=\fISTR\fR
select records with fields containing a string.
.TP
\fB\-n\fR, \fB\-\-number\fR=\fINUM\fR,...
select specific records by position, with ranges.
.TP
\fB\-m\fR, \fB\-\-random\fR=\fINUM\fR
select a given number of random records.
.SS "Field selection options:"
.TP
\fB\-f\fR, \fB\-\-fields\fR=\fIFIELDS\fR
comma\-separated list of field names with optional
subscripts.
.SS "Actions:"
.TP
\fB\-s\fR, \fB\-\-set\fR=\fIVALUE\fR
change the value of the selected fields.
.TP
\fB\-a\fR, \fB\-\-add\fR=\fIVALUE\fR
add the selected fields with the given value.
.TP
\fB\-S\fR, \fB\-\-set\-add\fR=\fIVALUE\fR
change the value of the selected fields.  If they don't
exist then add a new field with that value.
.TP
\fB\-r\fR, \fB\-\-rename\fR=\fINAME\fR
rename the selected fields to a given name.  If an entire
record set is selected then the field is renamed in the
record descriptor as well.
.TP
\fB\-d\fR, \fB\-\-delete\fR
delete the selected fields.
.TP
\fB\-c\fR, \fB\-\-comment\fR
comment out the selected fields.
.SH AUTHOR
Written by Jose E. Marchesi.
.SH "REPORTING BUGS"
Report bugs to: bug\-recutils@gnu.org
.br
GNU recutils home page: <http://www.gnu.org/software/recutils/>
.br
General help using GNU software: <http://www.gnu.org/gethelp/>
.SH COPYRIGHT
Copyright \(co 2010, 2011, 2012, 2013, 2014 Jose E. Marchesi.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B recset
is maintained as a Texinfo manual.  If the
.B info
and
.B recset
programs are properly installed at your site, the command
.IP
.B info recutils
.PP
should give you access to the complete manual.


NEWS
This NEWS file records noteworthy changes, very tersely.
See the manual for detailed information.

  Copyright 2010-2014 Jose E. Marchesi.

  Copying and distribution of this file, with or without modification,
  are permitted in any medium without royalty provided the copyright
  notice and this notice are preserved.

 Version 1.7 (25 March 2014)
* Utilities:
  . rec2csv now supports a -d option to change the delimiter
    character in CSV files.
  . It is now possible to specify several recfiles in the
    command line containing anonymous records.
* Format:
  . Support for restricting the fields that can appear in records
    with the new %allowed special field.
  . If a tab separates a field name and a field value is now
    ignored and not considered part of the field value.
* librec:
  . The current version of the library is now 1.
* Emacs mode:
  . It now possible to jump several records while navigating
    using a prefix argument to the next/previous commands.
* Documentation:
  . The examples in the manual have been improved, and well as many
    other aspects: formatting, English, etc.  Special thanks to Karl
    Berry for pointing out the problems and suggesting solutions.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.6 (03 November 2013)
* Utilities:
  . Support for joins and foreign keys.
  . Grouping records by one or more fields.
  . Sorting recors by several fields.
  . Specifying rewrite rules in field expressions.
  . Writing field values in rows.
  . Dot-notation in selection expressions.
  . Support for the >= and <= operators in selection expressions.
  . Aggregated functions: sum, min, max, avg.
  . New bash loadable builtin `readrec'.
* Format:
  . New UUID field type.
  . User provided arbitrary constraints with %constraint.
  . Pre-defined constants MIN and MAX can now be used in ranges.
* librec:
  . New high-level API: rec_db_query, rec_db_insert, rec_db_delete and
    rec_db_set.
  . Documentation in the rec.h header file extended and improved.
* Emacs mode:
  . Texinfo manual describing the usage of rec-mode.
  . Support for a mode hook.
  . Comments are now supported in records.
  . add-change-log-entry in a recfile now does the right thing.
  . Commands for copying/killing fields and records.
  . Support for undo in navigation mode.
  . Field folding.
  . Editing methods for the several field types.
  . Statistics.
* Documentation:
  . The user manual has been restructured and expanded.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.5 (13 January 2012)
* The utilities will now ask interactively for a password if it was
  not provided with the -s command line option.  This avoids security
  problems related to shell history files.
* Support for octal and hexadecimal numbers has been added.  They can
  be used in both the records and selection expressions.
* It is now possible to select a given number of random records in
  many of the utilities using the -m command line option.
* The -n option now accepts a list of indexes, supporting ranges.
* The new -U (uniq) option for recsel removes duplicated fields in the
  output records.
* The new -q (quick) option allows to quickly search for the desired
  record without having to provide a complete selection expression.
* Auto generated fields are now considered integers by default.  This
  avoids repetitive patterns in record descriptors involving %auto and
  %type.
* Tab characters are now allowed in blank lines betwwen records.
* The API in rec.h is now better documented with comments, and
  improved.
* recfix now exits with an error status if there is a parse error in
  some input file.
* The usage of the internal data structures has been _vastly_
  improved, resulting in a much faster operation.
* Internal cleanup and code factorization.
* Many, many, many bug fixes :D

 Version 1.4 (03 November 2011)
* Support for encryption with the new %confidential special field.
* Support for named types with the new %typedef special field.
* Support for sorting with the new %sort special field.
* Recutils support for org-babel is now included in the distribution
  as ob-rec.el.  It allows to read rec data into org documents in both
  raw and table formats.
* It is now possible to use double quotes for delimiting string
  literals in selection expressions, as well as single quotes.
  Unescapted newline characters are also supported.
* The manual has been vastly improved.
* recfix now supports several operations: check, sort, encrypt and
  decrypt.
* recins now generates parseable locale independent dates when
  generating time-stamps in auto-fields.
* Lots of bug fixes.

 Version 1.3 (11 February 2011)
* A quite bad (and embarrassing) bug introduced in 1.2 has been
  fixed.  The bug was causing recins, recdel and recset to report
  random integrity failures, due to a uninitialized variable.
* New special field %size, that allows to define constraints in the
  size of record sets.
* Several fixes in the user manual.
* Little portability issue that prevented one of the tests to compile
  in macos x has been fixed.
* Translations updates.

 Version 1.2 (4 February 2011)
* New utility rec2csv, to export rec data to csv files.
* Support for auto-incremented fields and timestamps with the new
  %auto field in record descriptors.
* Support for comments in enum type descriptions.
* recset now supports a new "set or add field" operation, implementing
  the common case of setting the value of a field in case it exists,
  or else add it to the record.
* The flavor of regular expressions accepted in both selection
  expressions and type descriptions is now documented in the user
  manual.
* Translations for German, Finnish, Swedish and Dutch.
* recdel now performs an integrity check before to complete the
  deletion.
* The utilities no longer "expand" the contents of remote descriptors
  in the local ones.

utils/recset.c
/* -*- mode: C -*-
 *
 *       File:         recset.c
 *       Date:         Fri Apr  9 17:06:39 2010
 *
 *       GNU recutils - recset
 *
 */

/* Copyright (C) 2010-2013 Jose E. Marchesi */

/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>

#include <getopt.h>
#include <string.h>
#include <stdlib.h>
#include <xalloc.h>
#include <gettext.h>
#define _(str) gettext (str)

#include <rec.h>
#include <recutl.h>

/*
 * Forward prototypes.
 */

static void recset_parse_args (int argc, char **argv);
static void recset_process_actions (rec_db_t db);

/*
 * Global variables.
 */

char      *recutl_sex_str     = NULL;
rec_sex_t  recutl_sex         = NULL;
char      *recutl_quick_str   = NULL;
char      *recutl_fex_str     = NULL;
rec_fex_t  recutl_fex         = NULL;
char      *recutl_type        = NULL;
int        recset_action      = REC_SET_ACT_NONE;
char      *recset_value       = NULL;
char      *recset_new_field_name = NULL;
bool       recutl_insensitive = false;
char      *recset_file        = NULL;
bool       recset_force       = false;
bool       recset_verbose     = false;
bool       recset_external    = true;
bool       recset_descriptor_renamed = false;
size_t     recutl_random      = 0;

/*
 * Command line options management
 */

enum
  {
    COMMON_ARGS,
    RECORD_SELECTION_ARGS,
    FIELD_EXPR_ARG,
    ADD_ACTION_ARG,
    RENAME_ACTION_ARG,
    DELETE_ACTION_ARG,
    COMMENT_ACTION_ARG,
    SET_ACTION_ARG,
    SET_ADD_ACTION_ARG,
    FORCE_ARG,
    VERBOSE_ARG,
    NO_EXTERNAL_ARG
  };

static const struct option GNU_longOptions[] =
  {
    COMMON_LONG_ARGS,
    RECORD_SELECTION_LONG_ARGS,
    {"fields", required_argument, NULL, FIELD_EXPR_ARG},
    {"add", required_argument, NULL, ADD_ACTION_ARG},
    {"rename", required_argument, NULL, RENAME_ACTION_ARG},
    {"delete", no_argument, NULL, DELETE_ACTION_ARG},
    {"comment", no_argument, NULL, COMMENT_ACTION_ARG},
    {"set", required_argument, NULL, SET_ACTION_ARG},
    {"set-add", required_argument, NULL, SET_ADD_ACTION_ARG},
    {"force", no_argument, NULL, FORCE_ARG},
    {"verbose", no_argument, NULL, VERBOSE_ARG},
    {"no-external", no_argument, NULL, NO_EXTERNAL_ARG},
    {NULL, 0, NULL, 0}
  };

/*
 * Functions.
 */


void
recutl_print_help (void)
{
  /* TRANSLATORS: --help output, recset synopsis.
     no-wrap */
  printf (_("\
Usage: recset [OPTION]... [FILE]...\n"));

  /* TRANSLATORS: --help output, recset short description.
     no-wrap */
  fputs (_("\
Alter or delete fields in records.\n"), stdout);

  puts ("");
  /* TRANSLATORS: --help output, recset options.
     no-wrap */
  fputs (_("\
      --no-external                   don't use external descriptors.\n\
      --force                         alter the records even if violating record\n\
                                        restrictions.\n"), stdout);

  recutl_print_help_common ();

  puts ("");
  recutl_print_help_record_selection ();

  puts ("");
  /* TRANSLATORS: --help output, recset field selection options.
     no-wrap */
  fputs (_("\
Field selection options:\n\
  -f, --fields=FIELDS                 comma-separated list of field names with optional\n\
                                        subscripts.\n"), stdout);

  puts ("");
  /* TRANSLATORS: --help output, recset actions.
     no-wrap */
  fputs (_("\
Actions:\n\
  -s, --set=VALUE                     change the value of the selected fields.\n\
  -a, --add=VALUE                     add the selected fields with the given value.\n\
  -S, --set-add=VALUE                 change the value of the selected fields.  If they don't\n\
                                        exist then add a new field with that value.\n\
  -r, --rename=NAME                   rename the selected fields to a given name.  If an entire\n\
                                        record set is selected then the field is renamed in the\n\
                                        record descriptor as well.\n\
  -d, --delete                        delete the selected fields.\n\
  -c, --comment                       comment out the selected fields.\n"), stdout);

  puts ("");
  recutl_print_help_footer ();
}

/* This macro is used in recset_parse_args to avoid code repetition in
   the actions (set, add, delete, ...) switch cases..  */
#define CHECK_ACTION_PREREQ                     \
  do                                                            \
    {                                                           \
      if (!recutl_fex)                                          \
        {                                                       \
          recutl_fatal (_("please specify some field with -f.\n")); \
        }                                                       \
                                                                \
      if (recset_action != REC_SET_ACT_NONE)                     \
        {                                                       \
          recutl_fatal (_("please specify just one action.\n")); \
        }                                                       \
    }                                                           \
  while (0)

static void
recset_parse_args (int argc,
                   char **argv)
{
  int ret;
  char c;

  while ((ret = getopt_long (argc,
                             argv,
                             RECORD_SELECTION_SHORT_ARGS
                             "dct:s:S:a:f:r:",
                             GNU_longOptions,
                             NULL)) != -1)
    {
      c = ret;
      switch (c)
        {
          COMMON_ARGS_CASES
          RECORD_SELECTION_ARGS_CASES
        case FORCE_ARG:
          {
            recset_force = true;
            break;
          }
        case VERBOSE_ARG:
          {
            recset_verbose = true;
            break;
          }
        case FIELD_EXPR_ARG:
        case 'f':
          {
            recutl_fex_str = xstrdup (optarg);
            if (!rec_fex_check (recutl_fex_str, REC_FEX_SUBSCRIPTS))
              {
                recutl_fatal (_("invalid field expression in -f.\n"));
              }

            /* Create the field expression.  */
            recutl_fex = rec_fex_new (recutl_fex_str,
                                      REC_FEX_SUBSCRIPTS);
            if (!recutl_fex)
              {
                recutl_fatal (_("creating the field expression.\n"));
              }

            /* Sort it.  */
            rec_fex_sort (recutl_fex);

            break;
          }
        case SET_ACTION_ARG:
        case 's':
          {
            CHECK_ACTION_PREREQ;
            recset_action = REC_SET_ACT_SET;
            recset_value = xstrdup (optarg);
            break;
          }
        case RENAME_ACTION_ARG:
        case 'r':
          {
            CHECK_ACTION_PREREQ;
            if (rec_fex_size (recutl_fex) != 1)
              {
                recutl_fatal (_("the rename operation requires just one field with an optional subscript.\n"));
              }

            recset_action = REC_SET_ACT_RENAME;
            recset_value = xstrdup (optarg);

            /* Validate the new name.  */
            recset_new_field_name = recset_value;
            if (!rec_field_name_p (recset_new_field_name))
              {
                recutl_fatal (_("invalid field name %s.\n"), recset_value);
              }
            
            break;
          }
        case ADD_ACTION_ARG:
        case 'a':
          {
            CHECK_ACTION_PREREQ;
            recset_action = REC_SET_ACT_ADD;
            recset_value = xstrdup (optarg);
            break;
          }
        case SET_ADD_ACTION_ARG:
        case 'S':
          {
            CHECK_ACTION_PREREQ;
            recset_action = REC_SET_ACT_SETADD;
            recset_value = xstrdup (optarg);
            break;
          }
        case DELETE_ACTION_ARG:
        case 'd':
          {
            CHECK_ACTION_PREREQ;
            recset_action = REC_SET_ACT_DELETE;
            break;
          }
        case COMMENT_ACTION_ARG:
        case 'c':
          {
            CHECK_ACTION_PREREQ;
            recset_action = REC_SET_ACT_COMMENT;
            break;
          }
        case NO_EXTERNAL_ARG:
          {
            recset_external = false;
            break;
          }
        default:
          {
            exit (EXIT_FAILURE);
            break;
          }
        }
    }

  /* Read the name of the data source.  */
  if (optind < argc)
    {
      if ((argc - optind) != 1)
        {
          recutl_print_help ();
          exit (EXIT_FAILURE);
        }

      recset_file = argv[optind++];
    }

}

static void
recset_process_actions (rec_db_t db)
{
  int flags = 0;

  if (recutl_insensitive)
    {
      flags = flags | REC_F_ICASE;
    }

  if (!rec_db_set (db,
                   recutl_type,
                   recutl_index (),
                   recutl_sex,
                   recutl_quick_str,
                   recutl_random,
                   recutl_fex,
                   recset_action,
                   recset_value,
                   flags))
    recutl_out_of_memory ();
  
  /* Check the integrity of the resulting database.  */

  if (!recset_force && db)
    {
      recutl_check_integrity (db, recset_verbose, recset_external);
    }
}

int
main (int argc, char *argv[])
{
  rec_db_t db;
  
  recutl_init ("recset");

  /* Parse arguments.  */
  recset_parse_args (argc, argv);

  db = recutl_read_db_from_file (recset_file);
  if (!db)
    {
      recutl_fatal (_("cannot read file %s\n"), recset_file);
    }

  recset_process_actions (db);

  if (!recutl_file_is_writable (recset_file))
    {
      recutl_error (_("file %s is not writable.\n"), recset_file);
      return EXIT_FAILURE;
    }
  recutl_write_db_to_file (db, recset_file);

  return EXIT_SUCCESS;
}

/* End of recset.c */
